text/0000-futures.md

```diff
@@ -0,0 +1,811 @@
+/// notifications occur. Executors, such as thread pools, allow tasks to be
+/// spawned and are responsible for putting tasks onto ready queues when
+/// they are woken up, and polling them when they are ready.
+pub trait BoxExecutor {
+    /// Spawn the given task, polling it until completion.
+    ///
+    /// # Errors
+    ///
+    /// The executor may be unable to spawn tasks, either because it has
+    /// been shut down or is resource-constrained.
+    fn spawn(&mut self, task: Task) -> Result<(), SpawnError>;
+
+    /// Determine whether the executor is able to spawn new tasks.
+    ///
+    /// # Returns
+    ///
+    /// An `Ok` return means the executor is *likely* (but not guaranteed)
+    /// to accept a subsequent spawn attempt. Likewise, an `Err` return
+    /// means that `spawn` is likely, but not guaranteed, to yield an error.
+    fn status(&self) -> Result<(), SpawnError> {
+        Ok(())
+    }
+}
+
+pub struct Task { .. }
+
+// this impl is in `std` only:
+impl From<Box<dyn Future<Item = ()> + Send>> for Task { .. }
+
+/// Provides the reason that an executor was unable to spawn.
+pub struct SpawnError { .. }
+
+impl SpawnError {
+    /// Spawning is failing because the executor has been shut down.
+    pub fn shutdown() -> SpawnError;
+
+    /// Check whether this error is the `shutdown` error.
+    pub fn is_shutdown(&self) -> bool;
+
+    // additional error variants added over time...
+}
+```
+
+We need the executor trait to be usable as a trait object, which is why `Task`
+is constructed here from a boxed future. (In the no_std section, we'll see
+another constructor). In the long run, though, once we can take `dyn` by value,
+we would deprecate `BoxExecutor` and have:
+
+```rust
+trait Executor {
+    fn spawn(&mut self, task: Future<Item = ()> + Send) -> Result<(), SpawnError>;
+    fn status(&self) -> Result<(), SpawnError> { .. }
+}
+
+impl<E: BoxExecutor> Executor for E {
+    /* implement by boxing */
+}
+```
+
+This is why the RFC proposes the name `BoxExecutor` for the trait.
+
+In addition to the above, the `core::task` module will include the following API
+for helping detect bugs:
+
+```rust
+/// Marks the current thread as being within the dynamic extent of an
+/// executor.
+///
+/// Executor implementations should call this function before beginning to
+/// execute a tasks, and drop the returned `Enter` value after completing
+/// task execution:
+///
+/// ```rust
+/// let enter = enter().expect("...");
+/// /* run task */
+/// drop(enter);
+/// ```
+///
+/// Doing so ensures that executors aren't accidentally invoked in a nested fashion.
+///
+/// # Error
+///
+/// Returns an error if the current thread is already marked, in which case the
+/// caller should panic with a tailored error message.
+pub fn enter() -> Result<Enter, EnterError>
+```
+
+As stated in the doc comment, the expectation is that all executors will wrap
+their task execution within an `enter` to detect inadvertent nesting.
```

Comment:
This API grew from our experience of users using a blocking executor *inside* a future that is itself being executed on some blocking executor. In most cases, the app would just deadlock, since the `future.wait()` would call park the thread until it was ready, when the only way it could know it was ready was `epoll` on the same thread.

If we could detect blocking operations of all kinds while in the context of an `Enter` guard, that'd be **amazing!**

Reply:
Here is an attempt at minimally phrasing this rationale in the doc comments.

"Doing so ensures that executors aren't accidentally invoked in a nested fashion. When that happens, the inner executor can block waiting for an event that can only be triggered by the outer executor, leading to a deadlock."

(Also, I would also love for OSes to stop overusing blocking APIs so much. When even reading from or writing to a memory address is a potentially blocking operation, it gets kind of ridiculous...)